### PR TITLE
Bugfix in agentset to handle addition and removal correctly

### PR DESCRIPTION
**Pull request by quaquel**
Fix for #1959 and additional unittests that test specifically on this

### CODE REVIEW

File: mesa/agent.py [hunk 1]
```diff
@@ -251,7 +251,8 @@ def do(
         Returns:
             AgentSet | list[Any]: The results of the method calls if return_results is True, otherwise the AgentSet itself.
         """
-        res = [getattr(agent, method_name)(*args, **kwargs) for agent in self._agents]
+        # we iterate over the actual weakref keys and check if weakref is alive before calling the method
+        res = [getattr(agent(), method_name)(*args, **kwargs) for agent in self._agents.keyrefs() if agent()]
```
Comment (EwoutH) [inline mesa/agent.py:255]: Does this need to be a list comprehension `[]`, or can ir also be a regular comphrension `()`, and that we return `list(res)` if requested.

Comment (quaquel) [reply mesa/agent.py:255]: No, it most be a list comprehension because otherwise it becomes a generator and the method does not get executed.

Comment (EwoutH) [reply mesa/agent.py:255]: Right, good to know

Comment (rht) [inline mesa/agent.py:255]: I think it'd be clearer to name `agent_ref` instead of `agent`. Because the call `agent_ref()` is the actual agent themselves.

**Comment by Corvince** [comment]
pre-commit.ci autofix

**Review by Corvince — APPROVED** [review]
Seems good to me! Thanks for investigating. I must admit I haven't wrapped my head around weakrefs completely. This error went a bit above my head

**Comment by quaquel** [comment]
It was a bit of a learning experience for me as well. I am familiar with weakrefs from Java so that helped. 

By the way, the code could  have been even more elegant by using the walrus operator (first time I came across a reason to use it).

```python
res = [getattr(agent, method_name)(*args, **kwargs) for agentref in self._agents.keyrefs() if (agent := agentref())]
```

**Comment by Corvince** [comment]
> By the way, the code could  have been even more elegant by using the walrus operator (first time I came across a reason to use it).
> 

That's true! I actually stumbled a bit while reading the code thinking about agentref being executed twice. But I didn't think of the walrus! But it's actually used in the space code for a similar list comprehension in get_cell_list_contents

**Comment by quaquel** [comment]
Should this result in a 2.2.1 release?

**Comment by EwoutH** [comment]
In my opinion, yes.

**Comment by EwoutH** [comment]
To inform you, we're currently blocked on our PyPI pipeline being broken (see https://github.com/projectmesa/mesa/discussions/1962). I was supposed to meet with @jackiekazil to resolve that issue, but I unfortunately missed that meeting. Working on it.

**Comment by Corvince** [comment]
There was another "bug" we should fix first imho. `Scheduler.agents_by_type` changed its return type, which is a breaking change. This should be reverted for 2.2 and then we should release a bugfix release.

**Comment by quaquel** [comment]
> There was another "bug" we should fix first imho. `Scheduler.agents_by_type` changed its return type, which is a breaking change. This should be reverted for 2.2 and then we should release a bugfix release.

To make sure I understand: `RandomActivationByType.agents_by_type` is now a `[type, AgentSet]` but it should have been `[type, [agent][int]]`? So, a simple property solution is all that is needed, and moving the current `self.agents_by_type` to `self._agents_by_type`.


```python
@property
def agents_by_type(self):
	agentsbytype = defauldict(dict)
	for k, v in self._agents_by_type.items():
		agentsbytype[k] = {agent:agent.unique_id for agent in v}

	return agentsbytype

```

**Comment by Corvince** [comment]
Yes, but you might add a warning that the return type is deprecated and mesa 3.0 will return an Agentset (which imho makes much more sense)

**Comment by EwoutH** [comment]
> add a warning that the return type is deprecated and mesa 3.0 will return an Agentset

Agreed

**Pull request merged**